File: movie-infra/src/main/java/org/bailey/repository/MovieRepositoryImpl.java
package org.bailey.repository;

import org.bailey.domain.entity.Movie;
import org.bailey.domain.repository.MovieRepository;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;

@Repository
public class MovieRepositoryImpl implements MovieRepository {

    private final JpaMovieRepository jpa; // ✅ domain 레이어만 의존

    public MovieRepositoryImpl(JpaMovieRepository jpa) {
        this.jpa = jpa;
    }

    @Override
    public List<Movie> findAll() {
        return jpa.findAll();
    }

    @Override
    public Optional<Movie> findById(Long id) {
        return jpa.findById(id);
    }
}

File: infrastructure/mysql/src/main/kotlin/com/hanghe/redis/movie/MovieRepositoryImpl.kt
package com.hanghe.redis.movie

import com.querydsl.core.types.dsl.BooleanExpression
import com.querydsl.jpa.impl.JPAQueryFactory
import org.springframework.stereotype.Repository

@Repository
class MovieRepositoryImpl(
    private val movieJpaRepository: MovieJpaRepository,
    private val jpaQueryFactory: JPAQueryFactory
) : MovieRepository {

    override fun findAll(): List<MovieEntity> {
        return movieJpaRepository.findAll()
    }

    override fun search(title: String?, genre: String?): List<MovieEntity> {
        return jpaQueryFactory.selectFrom(qMovie)
            .where(
                likeTitle(title),
                eqGenre(genre)
            )
            .orderBy(qMovie.releaseDate.desc())
            .fetch()
    }

    private fun likeTitle(title: String?): BooleanExpression? {
        return title?.let { qMovie.title.like("%$it%") }
    }

    private fun eqGenre(genre: String?): BooleanExpression? {
        return genre?.let { qMovie.genre.eq(MovieGenre.valueOf(it)) }
    }

    companion object {
        private val qMovie = QMovieEntity.movieEntity
    }

}
